src/main/java/com/redhat/devtools/lsp4ij/features/navigation/LSPGotoDeclarationHandler.java

```diff
@@ -42,25 +47,53 @@ public class LSPGotoDeclarationHandler implements GotoDeclarationHandler {
 
     @Nullable
     @Override
-    public PsiElement[] getGotoDeclarationTargets(@Nullable PsiElement sourceElement, int offset, Editor editor) {
+    public PsiElement[] getGotoDeclarationTargets(@Nullable PsiElement sourceElement,
+                                                  int offset,
+                                                  Editor editor) {
         Project project = editor.getProject();
         if (project == null || project.isDisposed()) {
             return PsiElement.EMPTY_ARRAY;
         }
-        PsiFile psiFile = sourceElement.getContainingFile();
+        PsiFile psiFile = sourceElement != null ? sourceElement.getContainingFile() : null;
         if (psiFile == null) {
             return PsiElement.EMPTY_ARRAY;
         }
         if (!LanguageServersRegistry.getInstance().isFileSupported(psiFile)) {
             return PsiElement.EMPTY_ARRAY;
         }
+
+        // If this was called for a populated semantic tokens view provider, just get the target directly from it
+        if (psiFile.getViewProvider() instanceof LSPSemanticTokensFileViewProvider semanticTokensFileViewProvider) {
```

Comment:
Pretty much what the comment above say, but if this is a semantic tokens-backed file, we can see if the element at the specified offset is a reference and, if so, we can resolve that reference to get the target element(s).